### PR TITLE
Fix crafter

### LINKED ISSUE #5322: Workers keep unnecessary item in their inventory
minecolonies-0.13.10-ALPHA-universal.jar
structurize-0.13.4-ALPHA.jar

This is the healer personal inventory :

and here is one builder's rack :

### LINKED ISSUE #5350: Builder places none to at most single digits worth of blocks before going to get more.
Version: 
minecolonies-0.13.14-ALPHA-universal.jar
structurize-0.13.4-ALPHA.jar
### Expected behavior

The builder places everything they bring with them before heading back for resources.

### Actual behaviour

They only place a few blocks and then run back for resources even though they could be placing more.

https://youtu.be/hKKBIszK1XE
### Steps to reproduce the problem
Outside of this only started when I went to upgrade the fortress university to tier 3 I have no clue whatsoever as to how to replicate this.

### Logs: Latest.log and crashlog at https://gist.github.com/
https://gist.github.com/jt1399/27e111f8339f132195552327ccd2cf52

### LINKED ISSUE #5351: Delivery Man Not Fulfilling Crafting Request For Builder
Version: 0.13.18 (Valhesia 2 version 2.2.3)

### Expected behavior
When the builder wants to craft something, the deliveryman should bring the materials the builder needs to craft with.

### Actual behaviour
My builder is currently requesting 15 dark oak logs so that he can craft dark oak planks.  He is standing in his hut waiting for them to appear.  Meanwhile, my warehouse has 300 dark oak logs in it, and my delivery man is standing in the warehouse doing nothing. Here is a picture of what my clipboard shows about the request:
Untitled

Also, oddly, the builder does not have a "I need 15 dark oak logs to do my job" request as he usually does when he requires a resource to build.  Instead he has the 'soft' version, "My job would be easier if I had 15 dark oak logs".

### Steps to reproduce the problem

1. Teach builder to make planks out of wood.
2. Build something that requires planks.
3. Watch the delivery man stand around doing nothing...

### LINKED ISSUE #5354: Workers working but not crafting anything
minecolonies-0.13.20-ALPHA-universal.jar
structurize-0.13.4-ALPHA.jar

I didn't check every crafter but the sawmill has the same issue.

https://youtu.be/gxbJuF7wcWM

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #5354
Closes #5351
Closes #5350
Closes #5322

# Changes proposed in this pull request:
- Fix crafrer issues
- Fix memory leak
-

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java [hunk 1]
```diff
@@ -1507,11 +1507,14 @@ public void onRequestedRequestComplete(@NotNull final IRequestManager manager, @
...
 
-        if (!getCompletedRequestsByCitizen().containsKey(citizenThatRequested))
+        if (citizenThatRequested != 0)
         {
-            getCompletedRequestsByCitizen().put(citizenThatRequested, new ArrayList<>());
+            if (!getCompletedRequestsByCitizen().containsKey(citizenThatRequested))
+            {
+                getCompletedRequestsByCitizen().put(citizenThatRequested, new ArrayList<>());
+            }
+            getCompletedRequestsByCitizen().get(citizenThatRequested).add(request.getId());
```
Comment (MatthiasMann) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java]: getCompletedRequestsByCitizen().computeIfAbsent(citizenThatRequested, ArrayList::new).add(request.getId());

Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java]: Good point, old code, but good point =D

**Review by someaddons — DISMISSED** [review]

File: src/api/java/com/minecolonies/api/util/InventoryUtils.java [hunk 2]
```diff
@@ -1776,21 +1776,16 @@ public static int transferXOfFirstSlotInItemHandlerWithIntoNextFreeSlotInItemHan
...
                 {
-                    sourceHandler.insertItem(desiredItemSlot, returnStack, false);
                     break;
                 }
                 // Only reduce if successfully inserted.
-                else
-                {
-                    currentAmount -= returnStack.getCount();
-                }
+                currentAmount -= returnStack.getCount();
```
Comment (someaddons) [inline src/api/java/com/minecolonies/api/util/InventoryUtils.java:1794]: think this should only reduce if the addItemStackToItemHandler is successfull

Comment (Raycoms) [reply src/api/java/com/minecolonies/api/util/InventoryUtils.java:1794]: If not, it breaks the loop.

Comment (Raycoms) [reply src/api/java/com/minecolonies/api/util/InventoryUtils.java:1794]: if !transfer
break

That's why.

**Review by someaddons — APPROVED** [review]

**Pull request merged**